components/director/internal/tenantfetcher/service.go

```diff
@@ -38,44 +49,99 @@ type QueryConfig struct {
-	tenantStorageService TenantStorageService
-	providerName         string
-	fieldMapping         TenantFieldMapping
-
-	retryAttempts uint
+	queryConfig                     QueryConfig
+	transact                        persistence.Transactioner
+	kubeClient                      KubeClient
+	eventAPIClient                  EventAPIClient
+	tenantStorageService            TenantStorageService
+	runtimeStorageService           RuntimeStorageService
+	providerName                    string
+	fieldMapping                    TenantFieldMapping
+	movedRuntimeByLabelFieldMapping MovedRuntimeByLabelFieldMapping
+	labelDefService                 LabelDefinitionService
+	retryAttempts                   uint
+	movedRuntimeLabelKey            string
 }
 
-func NewService(queryConfig QueryConfig, transact persistence.Transactioner, kubeClient KubeClient, fieldMapping TenantFieldMapping, providerName string, client EventAPIClient, tenantStorageService TenantStorageService) *Service {
+func NewService(queryConfig QueryConfig, transact persistence.Transactioner, kubeClient KubeClient, fieldMapping TenantFieldMapping, movRuntime MovedRuntimeByLabelFieldMapping, providerName string, client EventAPIClient, tenantStorageService TenantStorageService, runtimeStorageService RuntimeStorageService, labelDefService LabelDefinitionService, movedRuntimeLabelKey string) *Service {
 	return &Service{
-		transact:             transact,
-		kubeClient:           kubeClient,
-		fieldMapping:         fieldMapping,
-		providerName:         providerName,
-		eventAPIClient:       client,
-		tenantStorageService: tenantStorageService,
-		queryConfig:          queryConfig,
+		transact:                        transact,
+		kubeClient:                      kubeClient,
+		fieldMapping:                    fieldMapping,
+		providerName:                    providerName,
+		eventAPIClient:                  client,
+		tenantStorageService:            tenantStorageService,
+		runtimeStorageService:           runtimeStorageService,
+		queryConfig:                     queryConfig,
+		movedRuntimeByLabelFieldMapping: movRuntime,
+		retryAttempts:                   retryAttempts,
+		labelDefService:                 labelDefService,
+		movedRuntimeLabelKey:            movedRuntimeLabelKey,
+	}
+}
+
+func (supplier *currentTenantsSupplier) Get(ctx context.Context) (map[string]bool, error) {
+	if supplier.currentTenantsMap != nil {
+		return supplier.currentTenantsMap, nil
+	}
+
+	currentTenants, listErr := supplier.tenantStorageService.List(ctx)
+	if listErr != nil {
+		return nil, errors.Wrap(listErr, "while listing tenants")
+	}
 
-		retryAttempts: retryAttempts,
+	supplier.currentTenantsMap = make(map[string]bool)
+	for _, ct := range currentTenants {
+		supplier.currentTenantsMap[ct.ExternalTenant] = true
 	}
+
+	return supplier.currentTenantsMap, nil
+}
+
+func (a syncTenantsAction) Execute() error {
+	if err := a.fn(); err != nil {
+		return errors.Wrap(err, "while "+a.name)
```

Comment:
Either remove "while" from here or from the actions as it currently puts 2x "while"